org/mozilla/jss/tests/JSSProvider.java

```diff
@@ -0,0 +1,85 @@
+        assert(item != null);
+        assert(item.startsWith(BASE_PACKAGE));
+        assert(item.endsWith("." + _type));
+    }
+
+    public static void shouldNotProvide(Provider p, String algo) {
+        assert(p.get(algo) == null);
+    }
+
+    public static void main(String[] args) throws Exception {
+        Provider p = Security.getProvider("Mozilla-JSS");
+        System.out.println(p);
+
+        CryptoManager.initialize(args[0]);
+        CryptoManager cm = CryptoManager.getInstance();
+        cm.setPasswordCallback(new FilePasswordCallback(args[1]));
+
+        p = Security.getProvider("Mozilla-JSS");
+        System.out.println(p);
+
+        // Validate assumptions about how provider.get() works:
+        //
+        // Something which by no means should ever be implemented returns
+        // null.
+        shouldNotProvide(p, "Charlie's Chocolate Factory");
+
+        // Validate that our provider implements certain interfaces we care
+        // about. In particular, ensure they're overriden as part of our
+        // org.mozilla.jss.provider package, and that the class name matches
+        // what we expect our implementation to be.
+        shouldProvide(p, "KeyManagerFactory.NssX509", "JSSKeyManagerFactory");
+
+        // Validate that our provider is "default"; that is, when we get an
+        // algorithm instance without explicitly providing the provider, we
+        // end up with an instance from our provider.
+        Mac m = Mac.getInstance("HmacSHA512");
+        assert(m.getProvider().getName() == p.getName());
+
+        // Our KeyManagerFactory and TrustMangerFactory should return KeyManagers
+        // and TrustManagers from our class namespace.
+        KeyManagerFactory kmf = KeyManagerFactory.getInstance("NssX509");
+        assert(kmf.getKeyManagers().length > 0);
+        for (KeyManager km : kmf.getKeyManagers()) {
+            System.out.println("got KeyManager: " + km);
+            assert(km instanceof org.mozilla.jss.provider.javax.crypto.JSSKeyManager);
+        }
+
+        TrustManagerFactory tmf = TrustManagerFactory.getInstance("NssX509");
+        assert(tmf.getTrustManagers().length > 0);
+        for (TrustManager tm : tmf.getTrustManagers()) {
+            System.out.println("got TrustManager: " + tm);
+            assert(tm instanceof org.mozilla.jss.provider.javax.crypto.JSSTrustManager);
+        }
+
+        // Our KeyStore should return instances or extensions of PK11Cert.
+        KeyStore ks = KeyStore.getInstance("PKCS11");
+        ks.load(null, null);
+        Enumeration<String> aliases = ks.aliases();
+        while (aliases.hasMoreElements()) {
+            String alias = aliases.nextElement();
+            java.security.cert.Certificate cert = ks.getCertificate(alias);
+
+            if (cert == null || cert.getClass() == null) {
+                continue;
+            }
+
+            assert(org.mozilla.jss.pkcs11.PK11Cert.class.isAssignableFrom(cert.getClass()));
```

Comment:
Would this work?
```
assert(cert instanceof org.mozilla.jss.pkcs11.PKCS11Cert);
```

Reply:
No, because there's derivative classes/interfaces and such that make this messy and include things that aren't an instance of `PKCS11Cert`. (If I recall correctly, this has to do with the JSS X509Certificate interface). 

What this is validating is that we can make a statement such as `PK11Cert something = cert`.

Reply:
On second thought, I think either is fine, I'll rewrite this to use `instanceof`.